### PR TITLE
Improve  performance and fix stability in CSR kernel functions

### PR DESCRIPTION
**Pull request by PetrovKP**
1. Fix problem with stability for sparse and dense cases
2. Improve performance up to x8 on sparse case on training

### CODE REVIEW

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb763b0f-6e0a-f1c6-b33e-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb76786f-edca-f1f5-b33e-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb76d37b-fe6a-f19f-8ba8-8cdcd4b723a1

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb76e94b-0358-f14c-9bd0-8cdcd4b723a1

**Comment by PetrovKP** [comment]
/intelci: run
set infra_branch=dev/isamsono/integrate_dal4py

**Comment by PetrovKP** [comment]
/intelci: run
set infra_branch=dev/isamsono/integrate_dal4py

**Comment by PetrovKP** [comment]
http://intel-ci.intel.com/eb76f1fb-f5d4-f1eb-8014-8cdcd4b723a1

**Comment by PetrovKP** [comment]
@Mergifyio rebase

**Comment by PetrovKP** [comment]
PCM with daal4py: http://intel-ci.intel.com/eb76f324-759c-f176-ae13-8cdcd4b723a1

**Comment by PetrovKP** [comment]
PCM with daal4py: http://intel-ci.intel.com/eb770ae1-57ae-f12c-b655-8cdcd4b723a1

**Comment by PetrovKP** [comment]
@Mergifyio rebase

**Comment by PetrovKP** [comment]
PCM with daal4py: http://intel-ci.intel.com/eb795c20-991e-f1d4-ab0e-8cdcd4b6d2c4

**Comment by PetrovKP** [comment]
PCM with daal4py: http://intel-ci.intel.com/eb796e3a-f85a-f145-9921-8cdcd4b6d2c4

**Comment by PetrovKP** [comment]
PCM with daal4py: http://intel-ci.intel.com/eb79c34f-13c0-f155-8c6a-8cdcd4b723a1

File: cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i [hunk 1]
```diff
@@ -148,40 +149,101 @@ services::Status KernelImplLinear<fastCSR, algorithmFPType, cpu>::computeInterna
...
+    const size_t nBlocks2  = nVectors2 / blockSize + !!(nVectors2 % blockSize);
 
-        if (k != (algorithmFPType)1.0 || b != (algorithmFPType)0.0)
+    const bool isSOARes = r->getDataLayout() & NumericTableIface::soa;
+
+    TlsMem<algorithmFPType, cpu> tlsMklBuff(blockSize * blockSize);
+    SafeStatus safeStat;
+    daal::conditional_threader_for((nVectors1 >= blockSize * 2), nBlocks1, [&, isSOARes](const size_t iBlock1) {
+        const size_t nRowsInBlock1 = (iBlock1 != nBlocks1 - 1) ? blockSize : nVectors1 - iBlock1 * blockSize;
+        const size_t startRow1     = iBlock1 * blockSize;
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i:165]:
```cpp
const size_t nRowsInBlock1 = min(startRow1 + blockSize, nVectors1) - startRow1;
```

The same below.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i [hunk 2]
```diff
@@ -148,40 +149,101 @@ services::Status KernelImplLinear<fastCSR, algorithmFPType, cpu>::computeInterna
...
+
+        WriteOnlyRows<algorithmFPType, cpu> mtRRows;
+        if (!isSOARes)
         {
-            daal::threader_for_optional(nVectors1, nVectors1, [=](size_t i) {
-                for (size_t j = 0; j < nVectors2; j++)
+            mtRRows.set(r, startRow1, nRowsInBlock1);
+            DAAL_CHECK_MALLOC_THR(mtRRows.get());
+        }
+        daal::conditional_threader_for((nVectors2 >= blockSize * 2), nBlocks2, [&, nVectors2, nBlocks2](const size_t iBlock2) {
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i:179]:
```cpp
(nVectors2 >= blockSize * 2)
-> 
nBlocks2 >= 2
?
```

The same below.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i [hunk 3]
```diff
@@ -148,40 +149,101 @@ services::Status KernelImplLinear<fastCSR, algorithmFPType, cpu>::computeInterna
...
+
+            if (!isSOARes)
+            {
+                algorithmFPType * const dataR = mtRRows.get();
+
+                const size_t ldc = nVectors2;
+                SpBlas<algorithmFPType, cpu>::xgemm_a_bt(dataA1, colIndicesA1, rowOffsetsA1, dataA2, colIndicesA2, rowOffsetsA2, nRowsInBlock1,
+                                                         nRowsInBlock2, nFeatures, dataR + startRow2, ldc);
+
+                if (k != (algorithmFPType)1.0 || b != (algorithmFPType)0.0)
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i:198]: C casts -> C++ casts

The same below.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i [hunk 4]
```diff
@@ -168,7 +159,17 @@ services::Status KernelImplRBF<fastCSR, algorithmFPType, cpu>::computeInternalMa
...
+        const size_t * colIndicesA1    = mtA1.cols();
+        const size_t * rowOffsetsA1    = mtA1.rows();
+
+        WriteOnlyRows<algorithmFPType, cpu> mtR(r, 0, nVectors1);
+        DAAL_CHECK_BLOCK_STATUS(mtR);
+        algorithmFPType * dataR = mtR.get();
+
+        SpBlas<algorithmFPType, cpu>::xsyrk_a_at(dataA1, colIndicesA1, rowOffsetsA1, nVectors1, a1->getNumberOfColumns(), dataR, nVectors2);
 
         daal::threader_for_optional(nVectors1, nVectors1, [=](size_t i) {
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i:174]: Why optional not just simple parallel for?

File: cpp/daal/src/externals/service_spblas.h [hunk 5]
```diff
@@ -201,15 +201,15 @@ struct SpBlas
 
         splitCSR2CSC(a, ja, ia, n, nRowsInCommonBlock, nRowsInTailBlock, nBlocks, valuesCSC, colIdxCSC, rowIdxCSC);
 
-        daal::threader_for(nBlocks * nBlocks, nBlocks * nBlocks, [=](size_t idx) {
+        daal::conditional_threader_for((m * m > 512 * 512), nBlocks * nBlocks, [=](size_t idx) {
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/externals/service_spblas.h:204]:
```cpp
(m * m > 512 * 512)
->
m > 512
```

The same below.

File: cpp/daal/src/externals/service_spblas.h [hunk 6]
```diff
@@ -178,7 +178,7 @@ struct SpBlas
     }
 
 public:
-    static services::Status xsyrk_a_at(const fpType * a, const size_t * ja, const size_t * ia, size_t m, size_t n, fpType * c)
+    static services::Status xsyrk_a_at(const fpType * a, const size_t * ja, const size_t * ia, size_t m, size_t n, fpType * c, const size_t ldC)
     {
         size_t nBlocks = 50;
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/externals/service_spblas.h:183]: Strange hardcode :)

File: cpp/daal/src/externals/service_spblas.h [hunk 7]
```diff
@@ -236,7 +236,7 @@ struct SpBlas
                 PRAGMA_VECTOR_ALWAYS
                 for (size_t col = 0; col < cols; ++col)
                 {
-                    block_res.ptr[row * m + col] = 0.0;
+                    block_res.ptr[row * ldC + col] = 0.0;
```
Comment (SmirnovEgorRu) [inline cpp/daal/src/externals/service_spblas.h:239]: memset?

The same below.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i (hunk 1 above)
Comment (PetrovKP) [reply cpp/daal/src/algorithms/kernel_function/kernel_function_linear_csr_fast_impl.i:165]: In many places, this is exactly how it is done, if we change it - then everywhere.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i (hunk 4 above)
Comment (PetrovKP) [reply cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i:174]: this is a case for smo, some degradation is obtained in this case for a multiclass. We don't care too much about the performance of this case.

File: cpp/daal/src/externals/service_spblas.h (hunk 6 above)
Comment (PetrovKP) [reply cpp/daal/src/externals/service_spblas.h:183]: I would not touch :) My expectation is to replace it with MKL gemm.

File: cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i (hunk 4 above)
Comment (SmirnovEgorRu) [reply cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i:174]: If you don't care about perf of the case - better to make it simpler :)

**Review by SmirnovEgorRu — APPROVED** [review]

File: cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i (hunk 4 above)
Comment (PetrovKP) [reply cpp/daal/src/algorithms/kernel_function/kernel_function_rbf_csr_fast_impl.i:174]: Sure, but degradation for SMO is bad :(

**Pull request merged**